packages/mui-base/src/useList/listReducer.ts

```diff
@@ -0,0 +1,442 @@
+
+  let nextIndexCandidate: number;
+  let lookupDirection: 'next' | 'previous';
+  let wrapAround = !disableListWrap;
+
+  switch (offset) {
+    case 'reset':
+      if (defaultHighlightedIndex === -1) {
+        return null;
+      }
+
+      nextIndexCandidate = 0;
+      lookupDirection = 'next';
+      wrapAround = false;
+      break;
+
+    case 'start':
+      nextIndexCandidate = 0;
+      lookupDirection = 'next';
+      wrapAround = false;
+      break;
+
+    case 'end':
+      nextIndexCandidate = maxIndex;
+      lookupDirection = 'previous';
+      wrapAround = false;
+      break;
+
+    default: {
+      const newIndex = previouslyHighlightedIndex + offset;
+
+      if (newIndex < 0) {
+        if ((!wrapAround && previouslyHighlightedIndex !== -1) || Math.abs(offset) > 1) {
+          nextIndexCandidate = 0;
+          lookupDirection = 'next';
+        } else {
+          nextIndexCandidate = maxIndex;
+          lookupDirection = 'previous';
+        }
+      } else if (newIndex > maxIndex) {
+        if (!wrapAround || Math.abs(offset) > 1) {
+          nextIndexCandidate = maxIndex;
+          lookupDirection = 'previous';
+        } else {
+          nextIndexCandidate = 0;
+          lookupDirection = 'next';
+        }
+      } else {
+        nextIndexCandidate = newIndex;
+        lookupDirection = offset >= 0 ? 'next' : 'previous';
+      }
+    }
+  }
+
+  const nextIndex = findValidItemToHighlight(
+    nextIndexCandidate,
+    lookupDirection,
+    items,
+    disabledItemsFocusable,
+    isItemDisabled,
+    wrapAround,
+  );
+
+  return items[nextIndex] ?? null;
+}
+
+/**
+ * Toggles the selection of an item.
+ *
+ * @param item Item to toggle.
+ * @param selectedValues Already selected items.
+ * @param selectionLimit The number of items that can be simultanously selected. If null, there is no limit.
+ * @param itemComparer A custom item comparer function.
+ *
+ * @returns The new array of selected items.
+ */
+function toggleSelection<ItemValue>(
+  item: ItemValue,
+  selectedValues: ItemValue[],
+  selectionLimit: number | null,
+  itemComparer: (item1: ItemValue, item2: ItemValue) => boolean,
+) {
+  if (selectionLimit === 0) {
+    return [];
+  }
+
+  // Selection limit = 1 is a special case - we don't want to allow deselecting the item.
+  if (selectionLimit === 1) {
+    if (itemComparer(selectedValues[0], item)) {
+      return selectedValues;
+    }
+
+    return [item];
+  }
+
+  // The toggled item is selected; remove it from the selection.
+  if (selectedValues.some((sv) => itemComparer(sv, item))) {
+    return selectedValues.filter((sv) => !itemComparer(sv, item));
+  }
+
+  // The toggled item is not selected and the selected array is shorter than the limit - add to the selection.
+  if (selectionLimit === null || selectedValues.length < selectionLimit) {
+    return [...selectedValues, item];
+  }
+
+  // Truncate the selection to the limit (discard items with lower indexes).
```

Comment:
I see, but `selectionMode: 'none' | 'one' | 'multiple'` makes much more sense when reading the code, instead of figuring out that infinite is a valid number option for the multiple use-case. We can also name the options: `selectionMode: 'none' | 'single' | 'multiple'`

Reply:
OK, fair point. We don't need other options for now, plus it's an internal API anyway, so changing this when it's actually needed won't be a problem.